File: leetcode/cpp/implement-trie-prefix-tree.cpp
//
// Created by liangr on 9/1/17.
//

// https://leetcode.com/problems/implement-trie-prefix-tree

#include <string>
#include <vector>

using namespace std;

class TrieNode {
public:
    vector<TrieNode *> children;
    bool isWord;

    TrieNode(bool isWord = false) : isWord(isWord), children(26, NULL) {
    }
};

class Trie {
public:
    /** Initialize your data structure here. */
    Trie() : root(new TrieNode()) {
    }

    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode *node = root;
        for (auto ch: word) {
            int index = ch - 'a';
            if (node->children[index] == NULL) {
                node->children[index] = new TrieNode();
            }
            node = node->children[index];
        }
        node->isWord = true;
    }

    /** Returns if the word is in the trie. */
    bool search(string word) {
        TrieNode *node = find(word);
        return node != NULL && node->isWord;
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        TrieNode *node = find(prefix);
        return node != NULL;
    }

private:
    TrieNode *root;

    TrieNode* find(string key) {
        TrieNode *node = root;
        for (auto ch: key) {
            int index = ch - 'a';
            if (node->children[index] == NULL) {
                return NULL;
            }
            node = node->children[index];
        }
        return node;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * bool param_2 = obj.search(word);
 * bool param_3 = obj.startsWith(prefix);
 */

File: leetcode/implement_trie_prefix_tree.py
class TrieNode(object):
    def __init__(self):
        self.is_word = False
        self.children = {}

    def __repr__(self):
        return '{}. Children: {}.'.format(self.is_word, self.children)

class Trie(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        current = self.root
        for c in word:
            if c not in current.children:
                current.children[c] = TrieNode()
            current = current.children[c]
        current.is_word = True

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        current = self.root
        for c in word:
            if c not in current.children:
                return False
            current = current.children[c]
        return current.is_word

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        current = self.root
        for c in prefix:
            if c not in current.children:
                return False
            current = current.children[c]
        return True

if __name__ == '__main__':
    t = Trie()
    print(t.search('a'))
    print(t.startsWith('a'))
    t.insert('a')
    t.insert('abc')
    #t.insert('ab')

    print(t.search('ab'))
    print(t.search('abc'))

    print(t.startsWith('ab'))
    print(t.startsWith('abc'))
